File: Workpiece/Summoner/LocalPackages/com.gamelovers.mcp-unity/Editor/Tools/BatchExecuteTool.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Threading.Tasks;
using UnityEngine;
using UnityEditor;
using McpUnity.Unity;
using Newtonsoft.Json.Linq;
using Unity.EditorCoroutines.Editor;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for executing multiple operations in a single batch request.
    /// Supports sequential execution, stop-on-error, and atomic rollback.
    /// </summary>
    public class BatchExecuteTool : McpToolBase
    {
        private readonly McpUnityServer _server;

        public BatchExecuteTool(McpUnityServer server)
        {
            _server = server;
            Name = "batch_execute";
            Description = "Executes multiple tool operations in a single batch request. Reduces round-trips and enables atomic operations.";
            IsAsync = true;
        }

        public override void ExecuteAsync(JObject parameters, TaskCompletionSource<JObject> tcs)
        {
            EditorCoroutineUtility.StartCoroutineOwnerless(ExecuteBatchCoroutine(parameters, tcs));
        }

        private IEnumerator ExecuteBatchCoroutine(JObject parameters, TaskCompletionSource<JObject> tcs)
        {
            JArray operations = parameters["operations"] as JArray;
            bool stopOnError = parameters["stopOnError"]?.ToObject<bool?>() ?? true;
            bool atomic = parameters["atomic"]?.ToObject<bool?>() ?? false;

            // Validate operations array
            if (operations == null || operations.Count == 0)
            {
                tcs.SetResult(McpUnitySocketHandler.CreateErrorResponse(
                    "The 'operations' array is required and must contain at least one operation.",
                    "validation_error"
                ));
                yield break;
            }

            // Validate max operations (prevent abuse)
            if (operations.Count > 100)
            {
                tcs.SetResult(McpUnitySocketHandler.CreateErrorResponse(
                    "Maximum of 100 operations allowed per batch.",
                    "validation_error"
                ));
                yield break;
            }

            JArray results = new JArray();
            int succeeded = 0;
            int failed = 0;
            int undoGroup = -1;

            // Start undo group for atomic operations
            if (atomic)
            {
                Undo.IncrementCurrentGroup();
                undoGroup = Undo.GetCurrentGroup();
                Undo.SetCurrentGroupName("Batch Execute");
            }

            for (int i = 0; i < operations.Count; i++)
            {
                JObject operation = operations[i] as JObject;
                if (operation == null)
                {
                    results.Add(CreateOperationResult(i, null, false, null, "Invalid operation format"));
                    failed++;

                    if (stopOnError)
                    {
                        RevertIfAtomic(atomic, undoGroup);
                        break;
                    }
                    continue;
                }

                string toolName = operation["tool"]?.ToString();
                JObject toolParams = operation["params"] as JObject ?? new JObject();
                string operationId = operation["id"]?.ToString() ?? i.ToString();

                // Validate tool name
                if (string.IsNullOrEmpty(toolName))
                {
                    results.Add(CreateOperationResult(i, operationId, false, null, "Missing 'tool' name in operation"));
                    failed++;

                    if (stopOnError)
                    {
                        RevertIfAtomic(atomic, undoGroup);
                        break;
                    }
                    continue;
                }

                // Prevent recursive batch execution
                if (toolName == Name)
                {
                    results.Add(CreateOperationResult(i, operationId, false, null, "Cannot nest batch_execute operations"));
                    failed++;

                    if (stopOnError)
                    {
                        RevertIfAtomic(atomic, undoGroup);
                        break;
                    }
                    continue;
                }

                // Get the tool
                if (!_server.TryGetTool(toolName, out McpToolBase tool))
                {
                    results.Add(CreateOperationResult(i, operationId, false, null, $"Unknown tool: {toolName}"));
                    failed++;

                    if (stopOnError)
                    {
                        RevertIfAtomic(atomic, undoGroup);
                        break;
                    }
                    continue;
                }

                // Execute the tool
                JObject toolResult = null;
                Exception toolException = null;

                if (tool.IsAsync)
                {
                    var toolTcs = new TaskCompletionSource<JObject>();

                    try
                    {
                        tool.ExecuteAsync(toolParams, toolTcs);
                    }
                    catch (Exception ex)
                    {
                        toolException = ex;
                    }

                    // Wait for async tool completion (yield must be outside try-catch)
                    if (toolException == null)
                    {
                        while (!toolTcs.Task.IsCompleted)
                        {
                            yield return null;
                        }

                        if (toolTcs.Task.IsFaulted)
                        {
                            toolException = toolTcs.Task.Exception?.InnerException ?? toolTcs.Task.Exception;
                        }
                        else
                        {
                            toolResult = toolTcs.Task.Result;
                        }
                    }
                }
                else
                {
                    try
                    {
                        toolResult = tool.Execute(toolParams);
                    }
                    catch (Exception ex)
                    {
                        toolException = ex;
                    }
                }

                // Process result
                if (toolException != null)
                {
                    results.Add(CreateOperationResult(i, operationId, false, null, toolException.Message));
                    failed++;

                    if (stopOnError)
                    {
                        RevertIfAtomic(atomic, undoGroup);
                        break;
                    }
                }
                else if (toolResult != null)
                {
                    // Check if the result indicates an error
                    bool isError = toolResult["error"] != null;
                    bool isSuccess = toolResult["success"]?.ToObject<bool?>() ?? !isError;

                    if (isSuccess && !isError)
                    {
                        results.Add(CreateOperationResult(i, operationId, true, toolResult, null));
                        succeeded++;
                    }
                    else
                    {
                        string errorMessage = toolResult["error"]?["message"]?.ToString()
                            ?? toolResult["message"]?.ToString()
                            ?? "Tool execution failed";
                        results.Add(CreateOperationResult(i, operationId, false, toolResult, errorMessage));
                        failed++;

                        if (stopOnError)
                        {
                            RevertIfAtomic(atomic, undoGroup);
                            break;
                        }
                    }
                }
                else
                {
                    results.Add(CreateOperationResult(i, operationId, false, null, "Tool returned null result"));
                    failed++;

                    if (stopOnError)
                    {
                        RevertIfAtomic(atomic, undoGroup);
                        break;
                    }
                }

                // Yield to allow Unity to process other events
                yield return null;
            }

            // Collapse undo group
            if (atomic && undoGroup >= 0 && failed == 0)
            {
                Undo.CollapseUndoOperations(undoGroup);
            }

            // Build response
            string message;
            if (failed == 0)
            {
                message = $"Successfully executed {succeeded}/{operations.Count} operations.";
            }
            else if (atomic && stopOnError)
            {
                message = $"Batch execution failed and rolled back. {succeeded} operations succeeded before failure.";
            }
            else if (stopOnError)
            {
                message = $"Batch execution stopped on error. {succeeded}/{operations.Count} operations succeeded.";
            }
            else
            {
                message = $"Batch execution completed with errors. {succeeded}/{operations.Count} operations succeeded, {failed} failed.";
            }

            tcs.SetResult(new JObject
            {
                ["success"] = failed == 0,
                ["type"] = "text",
                ["message"] = message,
                ["results"] = results,
                ["summary"] = new JObject
                {
                    ["total"] = operations.Count,
                    ["succeeded"] = succeeded,
                    ["failed"] = failed,
                    ["executed"] = succeeded + failed
                }
            });
        }

        private void RevertIfAtomic(bool atomic, int undoGroup)
        {
            if (atomic && undoGroup >= 0)
            {
                Undo.RevertAllDownToGroup(undoGroup);
            }
        }

        private JObject CreateOperationResult(int index, string id, bool success, JObject result, string error)
        {
            var operationResult = new JObject
            {
                ["index"] = index,
                ["id"] = id ?? index.ToString(),
                ["success"] = success
            };

            if (success && result != null)
            {
                operationResult["result"] = result;
            }
            else if (!success)
            {
                operationResult["error"] = error ?? "Unknown error";
            }

            return operationResult;
        }
    }
}

File: Workpiece/Summoner/LocalPackages/com.gamelovers.mcp-unity/Server~/src/tools/batchExecuteTool.ts
import * as z from 'zod';
import { McpUnity } from '../unity/mcpUnity.js';
import { Logger } from '../utils/logger.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

const toolName = 'batch_execute';
const toolDescription = `Executes multiple tool operations in a single batch request.
Reduces network round-trips and enables atomic operations with rollback support.
Performance improvement: 10-100x for repetitive operations.`;

const operationSchema = z.object({
  tool: z.string().describe('The name of the tool to execute'),
  params: z.record(z.any()).optional().default({}).describe('Parameters to pass to the tool'),
  id: z.string().optional().describe('Optional identifier for this operation (for tracking in results)')
});

const paramsSchema = z.object({
  operations: z.array(operationSchema)
    .min(1, 'At least one operation is required')
    .max(100, 'Maximum of 100 operations allowed per batch')
    .describe('Array of operations to execute sequentially'),
  stopOnError: z.boolean()
    .default(true)
    .describe('If true, stops execution on the first error. Default: true'),
  atomic: z.boolean()
    .default(false)
    .describe('If true, rolls back all operations if any fails (uses Unity Undo system). Default: false')
});

/**
 * Result of a single operation in the batch
 */
interface OperationResult {
  index: number;
  id: string;
  success: boolean;
  result?: any;
  error?: string;
}

/**
 * Summary of batch execution
 */
interface BatchSummary {
  total: number;
  succeeded: number;
  failed: number;
  executed: number;
}

/**
 * Response from the batch execute tool
 */
interface BatchExecuteResponse {
  success: boolean;
  type: string;
  message: string;
  results: OperationResult[];
  summary: BatchSummary;
}

/**
 * Creates and registers the Batch Execute tool with the MCP server
 */
export function registerBatchExecuteTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, {
          operationCount: params.operations?.length,
          stopOnError: params.stopOnError,
          atomic: params.atomic
        });
        const result = await batchExecuteHandler(mcpUnity, params, logger);
        logger.info(`Tool execution completed: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function batchExecuteHandler(
  mcpUnity: McpUnity,
  params: z.infer<typeof paramsSchema>,
  logger: Logger
): Promise<CallToolResult> {
  // Validate operations array
  if (!params.operations || params.operations.length === 0) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "The 'operations' array is required and must contain at least one operation"
    );
  }

  if (params.operations.length > 100) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Maximum of 100 operations allowed per batch"
    );
  }

  // Validate no nested batch_execute operations
  for (const op of params.operations) {
    if (op.tool === 'batch_execute') {
      throw new McpUnityError(
        ErrorType.VALIDATION,
        "Cannot nest batch_execute operations"
      );
    }
  }

  logger.info(`Sending batch with ${params.operations.length} operations to Unity`);

  // Send the batch request to Unity
  const response = await mcpUnity.sendRequest({
    method: toolName,
    params: {
      operations: params.operations.map((op, index) => ({
        tool: op.tool,
        params: op.params ?? {},
        id: op.id ?? index.toString()
      })),
      stopOnError: params.stopOnError ?? true,
      atomic: params.atomic ?? false
    }
  }) as BatchExecuteResponse;

  // Format the response message
  let resultText = response.message || 'Batch execution completed';

  // Add summary details
  if (response.summary) {
    resultText += `\n\nSummary: ${response.summary.succeeded}/${response.summary.total} succeeded`;
    if (response.summary.failed > 0) {
      resultText += `, ${response.summary.failed} failed`;
    }
  }

  // Add individual results if there are failures or detailed info
  if (response.results && response.results.length > 0) {
    const failures = response.results.filter(r => !r.success);
    if (failures.length > 0) {
      resultText += '\n\nFailed operations:';
      for (const failure of failures) {
        resultText += `\n  - [${failure.id}] ${failure.error || 'Unknown error'}`;
      }
    }
  }

  // Determine if we should throw an error or return success
  if (!response.success && params.stopOnError) {
    // When stopOnError is true and we failed, throw to signal the error clearly
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      resultText
    );
  }

  return {
    content: [{
      type: 'text',
      text: resultText
    }]
  };
}
